docs/spec/overload.rst

```diff
@@ -99,8 +109,397 @@ variable is not sufficient.
+Refer to https://github.com/python/typing/issues/253#issuecomment-389262904
+for Ivan's explanation. I'm not convinced this exemption is necessary.
+Currently pyright copies the exemption. Do we want to codify this or leave it
+out?]
+
+If all arguments accepted by an overload are also always accepted by
+an earlier overload, the two overloads are said to "fully overlap".
+In this case, the latter overload will never be used. This condition
+is indicative of a programming error and should be reported by type
+checkers::
+
+  # These overloads fully overlap because the second overload
+  # accepts all arguments accepted by the first overload.
+
+  @overload
+  def func[T](x: T) -> T: ...
+  @overload
+  def func(x: int) -> int: ...
+
+
+Overload call evaluation
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+When a type checker evaluates the call of an overloaded function, it
+attempts to "match" the supplied arguments with one or more overloads.
+This section describes the algorithm that type checkers should use
+for overload matching.
+
+Only the overloads (the ``@overload``-decorated signatures) should be
+considered for matching purposes. The implementation, if provided,
+should be ignored for purposes of overload matching.
+
+
+Step 1: Examine the argument list to determine the number of
+positional and keyword arguments. Use this information to eliminate any
+overload candidates that are not plausible based on their
+input signatures.
+
+- If no candidate overloads remain, generate an error and stop.
+- If only one candidate overload remains, it is the winning match. Evaluate
+  it as if it were a non-overloaded function call and stop.
+- If two or more candidate overloads remain, proceed to step 2.
+
+
+Step 2: Evaluate each remaining overload as a regular (non-overloaded)
+call to determine whether it is compatible with the supplied
+argument list. Unlike step 1, this step considers the types of the parameters
+and arguments. During this step, do not generate any user-visible errors.
+Simply record which of the overloads result in evaluation errors.
+
+- If all overloads result in errors, proceed to step 3.
+- If only one overload evaluates without error, it is the winning match.
+  Evaluate it as if it were a non-overloaded function call and stop.
+- If two or more candidate overloads remain, proceed to step 4.
+
+
+Step 3: If step 2 produces errors for all overloads, perform
+"argument type expansion". Some types can be decomposed
+into two or more subtypes. For example, the type ``int | str`` can be
+expanded into ``int`` and ``str``.
+
+Expansion should be performed one argument at a time from left to
+right. Each expansion results in sets of effective argument types.
+For example, if there are two arguments whose types evaluate to
+``int | str`` and ``int | bytes``, expanding the first argument type
+results in two sets of argument types: ``(int, ?)`` and ``(str, ?)``.
+Here ``?`` represents an unexpanded argument type.
```

Comment:
The reason I used `?` is to emphasize that these are not expanded argument types, so they should be evaluated using normal evaluation rules. If the type checker uses inference context (as all static type checkers do), they should apply context in this case rather than using the type evaluated without context.

That may be getting too detailed and subtle for the spec given that inference context (aka bidirectional type inference) isn't mentioned anywhere else in the spec currently.

I'll remove the `?` because it has the potential to add confusion.